light-model-client/src/commonMain/kotlin/org/modelix/client/light/LightModelClient.kt

```diff
@@ -901,3 +907,66 @@ fun NodeData.asUpdateData(): NodeUpdateData {
 fun INode.isLoaded() = isValid
 fun <T : INode> Iterable<T>.filterLoaded() = filter { it.isLoaded() }
 fun <T : INode> Sequence<T>.filterLoaded() = filter { it.isLoaded() }
+
+data class NodeDataDependency(val client: LightModelClient, val id: NodeId) : IStateVariableReference<NodeData> {
+    override fun getGroup(): IStateVariableGroup {
+        return client.tryGetParentId(id)
+            ?.let { NodeDataDependency(client, it) }
+            ?: client
+    }
+
+    override fun read(): NodeData {
+        return client.getNode(id).getData()
+    }
+}
+
+private class NodesMap<V : Any>(val client: LightModelClient) {
+    private val map: MutableMap<NodeId, V> = HashMap()
+
+    operator fun get(key: NodeId): V? {
+        DependencyTracking.accessed(NodeDataDependency(client, key))
+        return map[key]
+    }
+
+    operator fun set(key: NodeId, value: V) {
+        if (map[key] == value) return
+        map[key] = value
+        DependencyTracking.modified(NodeDataDependency(client, key))
+    }
+
+    fun remove(key: NodeId): V? {
+        if (!map.containsKey(key)) return null
+        val result = map.remove(key)
+        DependencyTracking.modified(NodeDataDependency(client, key))
+        return result
+    }
+
+    fun clear() {
+        if (map.isEmpty()) return
+        val removedKeys = map.keys.toList()
+        map.clear()
+        for (key in removedKeys) {
+            DependencyTracking.modified(NodeDataDependency(client, key))
+        }
+    }
+
+    fun containsKey(key: NodeId): Boolean {
+        DependencyTracking.accessed(NodeDataDependency(client, key))
+        return map.containsKey(key)
+    }
+
+    fun isNotEmpty(): Boolean {
```

Comment:
`isNotEmpty` is called in `hasTemporaryIds`. 
Why do we not need to track access and modifications for it? / How is it used? / Might it be used in an incremental function?

```kotlin
    fun hasTemporaryIds(): Boolean = requiresRead {
        temporaryNodeAdapters.isNotEmpty() || nodesReferencingTemporaryIds.isNotEmpty()
    }
```

Reply:
Correct, but `hasTemporaryIds` isn't used anywhere. Just removed the dead code.